R/globals.R

```diff
@@ -25,6 +25,32 @@ register_s3_method <- function(pkg, generic, class, fun = NULL) {
   )
 }
 
+register_upgrade_message <- function(pkg, version) {
+  # Do nothing if the package isn't installed or
+  # meets the version requirement
+  if (system.file(package = pkg) == "") return(NULL)
+  if (packageVersion(pkg) >= as.character(version)) return(NULL)
+
+  msg <- sprintf(
+    "This version of Shiny is designed to work with '%s' >= %s.
+    Please upgrade via install.packages('%s').",
+    pkg, version, pkg
+  )
+
+  # emit startup message of the package is already loaded
+  if (pkg %in% loadedNamespaces()) {
+    packageStartupMessage(msg)
+  }
+
+  # Always register hook in case package is later unloaded & reloaded
```

Comment:
I think this comment needs a little more context so it doesn't throw us off in the future. The common case is that the package is loaded later (and not unloaded and reloaded). The fact that we're doing it unconditionally means that it will emit the message in case the package is unloaded and reloaded, but seems incidental, because it's much less common, and the user would have already seen the message.

Reply:
That makes sense. I copied that comment from `register_s3_method()`...do you want that updated as well?

Reply:
Yes, that would be good, thanks. `git blame` says that one is my fault :)